### ISSUE TITLE
WebID+TLS compatibility

### ISSUE THREAD

**Issue by elf-pavlik** [fkooman/indiecert#6]
http://webid.info/

* http://www.w3.org/2005/Incubator/webid/spec/tls/
* http://www.w3.org/ns/auth/cert#

1. Make sure that IndieCert generated certificates work with RPs
2. Make sure that non IndieCert generated WebID certificates work with IndieCert

@bblfish @melvincarvalho

**Comment by bblfish** [comment]
Do you have a pointer to the right IndieCert specification? And what is your take on this @elf-pavlik - you have been following the WebID work for 4 years or so and you know both worlds. How do you think they fit together? That will help save us some thinking time. ( Also for others: all the WebID specs are at http://webid.info/spec/ )

**Comment by fkooman** [comment]
IndieCert aims to be compatible with the IndieAuth protocol from the relying party point of view. There is not really much in terms of specification, not for IndieAuth and not for IndieCert. There is my blog post describing IndieCert at https://www.tuxed.net/fkooman/blog/indiecert.html, but it is already obsolete regarding "scope" requirements and distributed Indie{Cert,Auth}.

As for the certificates: currently they are generated and signed by a self signed CA to work "properly" with the web server checking if they were issued by the IndieCert CA. I'm thinking about allowing all client certificates to be valid for IndieCert (IndieCert only needs to be sure the browser has the private key belonging to the presented cert, and I think TLS assures that).

In addition it will be possible to include the user's homepage in the SAN field at some point to be compatible with WebID.

A fingerprint of the certificate is stored on the user's homepage, so not the complete certificate. See https://www.tuxed.net/fkooman/blog/indiecert.html#fingerprint for more information on that.

**Comment by elf-pavlik** [comment]
@bblfish IMO IndieCert server come very handy for people who just want to support IndieAuth and let it take care of all the complexity http://indiewebcamp.com/indieauth
e.g. I run NGINX and don't want to deal with requesting client certs etc. Instead I just delegate to some IndieAuth/IndieCert server, which gives me support for Mozilla Persona, Github, Twitter, Facebook etc. *and* those who use WebID+TLS can *also* use it :smile:

**Comment by melvincarvalho** [comment]
Good work.  As discussed on email

I'd love to see the dependency of indiecert.net removed from the flow.

Otherwise, looking forward to seeing this work used more.  Aligning one day with webid would also be a huge interoperability plus, but maybe something for v2.

**Comment by elf-pavlik** [comment]
@melvincarvalho have you noticed?
> As for the certificates: currently they are generated and signed by a self signed CA to work "properly" with the web server checking if they were issued by the IndieCert CA. I'm thinking about allowing all client certificates to be valid for IndieCert (IndieCert only needs to be sure the browser has the private key belonging to the presented cert, and I think TLS assures that).

**Comment by melvincarvalho** [comment]
@elf-pavlik I didnt see that no, and it sounds promising.  Id look forward to that feature, as it would maybe them be possible to use all the apps the linked data community build with this system and vice versa.

**Comment by bblfish** [comment]
Well it seems pretty easy. All you need to do is put a WebID in SAN of the cert - best do one with a #uri as shown in the diagram http://www.w3.org/2005/Incubator/webid/spec/identity/ - and to publish at the WebID profile a content negotiated document that can return profile information in html, json-ld, turtle, etc.. using the cert ontology. It is really easy for a browser to generate a key using keygen, as explained in the WebID-TLS document.

**Comment by elf-pavlik** [comment]
@bblfish let's don't get into # URIs please :-1: I would also propose to add optional support for RDFa sto WebID since many people will NOT do Content Negotiation and provide Turtle!

**Comment by bblfish** [comment]
ok, well if you want to slow down your server due to redirect problems then why do you bother asking me. And if content negotiation or formats holds you up then just grow up. please @elf-pavlik don't pull me into discussions with immature people.

**Comment by elf-pavlik** [comment]
@fkooman let's for now just make sure to include /foo#me etc. in SAN
@bblfish we'll try to get it working with @melvincarvalho RDFa WebID profile. IMO if you want to enforce Turte then you may face adoption problems out side of Semantic Web circles. But let's leave it here and just see how far we can get them aligned for now.

**Comment by bblfish** [comment]
Don't want to enforce turtle. I said you can offer a number of different serialisations.

**Comment by elf-pavlik** [comment]
You can also have Turtle in <script /> tag, similar as JSON-LD http://www.w3.org/TR/json-ld-syntax/#embedding-json-ld-in-html-documents

**Comment by fkooman** [comment]
For now I'll think about accepting WebID certs and not generate them. The user will just have to put the fingerprint of their WebID cert on their website with a `rel="me"` tag. Seems the easiest for now :) After that we can think about becoming more WebID compatible... content negotiation and thinking about fragments seems a bridge too far right now...

**Comment by elf-pavlik** [comment]
How to use rel="me" when using RDFa explained on http://indiewebcamp.com/indieauth#RDFa

**Comment by melvincarvalho** [comment]
sounds like a good first step, I could change my profile to do that

**Comment by melvincarvalho** [comment]
@elf-pavlik please dont allow your personal opinions on vanity uris and # uris to censor a productive discussion about interopeability.  I know you have strong views in this area, and I know you hold them sincerely.  But in the spirit of tolerance, let people have their say and make up their own minds.

**Comment by melvincarvalho** [comment]
you can have rel="me" and href = <fingerprint of cert> ni:/// (three slashes because no domain dependence) then the hash

**Comment by bblfish** [comment]
@fkooman producing WebIDs is much easier than authenticating people with WebID-TLS. WebID-TLS requires also hacking the TLS layer to bypass CA verification. You then also need to be able to parse Turtle, JSOn-LD and RDFa if you want to be very open, and be ready to do the verification step explained in the WebID-TLS step. Anyway: ideally you'd do both. But starting with producing WebIDs is a lot simpler.

**Comment by fkooman** [comment]
@bblfish I think I'll just only accept WebID certificates, should be relatively easy with `SSLVerifyClient optional_no_ca` in Apache, no hacking required. Then the fingerprint needs to be stored on the user's homepage and done :)

**Comment by melvincarvalho** [comment]
Sounds good!  Shout if you need help testing ...

**Comment by fkooman** [comment]
You can test it right now, indiecert.net accepts all certificates now. 

The only thing that still needs fixing is that right now `ni://indiecert.net/<fingerprint>` is needed on your homepage, so the domain must be there for now. I'll release a new version soonish where this is no longer required.

**Comment by fkooman** [comment]
@melvincarvalho were you able to test yet? I removed all domain references from the `ni:///` fingerprint stuff, should it should be usable in a generic way :-)

**Comment by melvincarvalho** [comment]
Internal Server Error

[curl] (#51) See http://curl.haxx.se/libcurl/c/libcurl-errors.html for an explanation of cURL errors [url] https://melvincarvalho.com/

Tho I havent yet put the fingerprint URI in my "me" is there an easy way to get that from a web service, do you know?

**Comment by fkooman** [comment]
@melvincarvalho ah, your certificate is configured wrong it seems :) That's why you get the 500 error. If you fix your cert you will see a page where you can copy/paste the fingerprint to your homepage :)

**Comment by melvincarvalho** [comment]
hmm configured wrong how?  Been using this cert for 5 years ... never had an issue ...

**Comment by fkooman** [comment]
@melvincarvalho the hostname of server doesn't match the one in the certificate, see https://www.ssllabs.com/ssltest/analyze.html?d=melvincarvalho.com&hideResults=on&latest

**Comment by melvincarvalho** [comment]
Thanks for the pointer. 

Hmmm my homepage is under http, not https ... im waiting for lets encrypt before making a change.  Does indie cert work with http too?  BTW from experience of testing TLS over the last few years about 90% of users used HTTP...

**Comment by fkooman** [comment]
Nope, it doesn't work with `http://` and it won't. You can use StartSSL or CAcert. Both work with IndieCert and are "free".

**Comment by melvincarvalho** [comment]
Thanks.  Will test again when I upgrade.

**Comment by fkooman** [comment]
```
            X509v3 Subject Alternative Name: critical
                URI:https://bob.example/profile#me
```

http://www.w3.org/2005/Incubator/webid/spec/tls/#the-certificate

issue #6

**Comment by fkooman** [comment]
So currently IndieCert does support WebID+TLS client certificates to authenticate, but does require you to put the fingerprint of your cert on your homepage using `<link rel="publickey" href="ni:///<fingerprint>">`. Following the flow on https://indiecert.net/ will tell you exactly what to put on your homepage. Generated certificates by IndieCert are not WebID+TLS compatible. I'd gladly point to a URL where people can generate a WebID+TLS compatible certificate! Ideally I'd remove the client certificate generation stuff from IndieCert completely!

**Comment by fkooman** [comment]
Adding a SAN to the certificate (for WebID+TLS support) is now implemented and available on https://indiecert.net/enroll. You can modify it at will. The validity of the provided URL will not be verified.
![screenshot from 2015-07-10 09-16-05](https://cloud.githubusercontent.com/assets/616885/8613891/53cc3cde-26e4-11e5-9b12-49d2bd3804e5.png)
![screenshot from 2015-07-10 09-15-44](https://cloud.githubusercontent.com/assets/616885/8613892/53d54d56-26e4-11e5-84ca-64eeaeb0953b.png)

**Comment by fkooman** [comment]
So next step is to verify the URL in the certificate and support "native" WebID+TLS for IndieAuth!
